File: ds/graph/lc_785_Is_Graph_Bipartite/java/Solution.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int[] color = new int[n];
        Arrays.fill(color, -1);

        for (int start = 0; start < n; start++) {
            if (color[start] == -1) {
                Stack<Integer> stack = new Stack();
                stack.push(start);
                color[start] = 0;

                while (!stack.isEmpty()) {
                    Integer node = stack.pop();
                    for (int nei : graph[node]) {
                        if (color[nei] == -1) {
                            stack.push(nei);
                            color[nei] = color[node] ^ 1;

                        }else if (color[nei] == color[node]){
                            return false;
                        }
                    }
                }
            }
        }
        return true;
    }
}

File: ds/graph/lc_785_Is_Graph_Bipartite/python/Solution.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        self.color = [0] * n
        for i in range(n):
            if self.color[i] == 0 and not self.colored(i, graph, 1):
                return False
        return True

    def colored(self, now, graph, c):
        self.color[now] = c
        for nei in graph[now]:
            if self.color[nei] == 0 and not self.colored(nei, graph, -c):
                return False
            elif self.color[nei] == c:
                return False
        return True